[browser] managed Diagnostics.Metrics and Diagnostics.Tracing single-threaded

pavelsavara opened the pull request:
This change enables Diagnostics.Metrics and Diagnostics.Tracing in single-threaded browser. 

The behavior is slightly different from desktop/server dotnet because the events are only propagated on the same thread as application logic. That's when the user code yields to the browser event loop. In well behaved applications, this is often, because that's how you allow the browser to re-render the page. If the user code is running long synchronous loop, the events would be only delivered later.

### Implementation

- excuded `CounterGroup` and `RuntimeEventSource` from browser build
- split `EventPipeEventDispatcher.Threads.cs` and `EventPipeEventDispatcher.Wasm.cs` - make it Task based
- updated `AggregationManager` - make it timer based for WASM

### Other

- fix nullable argument in `BeginInstrumentReporting`, `EndInstrumentReporting` and `InstrumentPublished`

### Tests

- enable `EventSourceSupport`, `MetricsSupport` MSBuild props for unit tests which need it
- enabled `WasmPerfTracing` for `ActivityTracking` test to work, because `EventPipeEventProvider.EventActivityIdControl` is in `libmono-component-diagnostics_tracing` which need to be linked in
- enable `TestSubscriptionManagerDisposal`
- enable `RuntimeMetricsTests`, `ActivityTracking`
- `MetricEventSourceTests` are `OuterLoop`, but I made them work locally by rewriting them to async.

Fixes https://github.com/dotnet/runtime/issues/93754
Contributes to https://github.com/dotnet/runtime/issues/76316

pavelsavara committed:
Implementation

- split AggregationManager.Threads.cs and AggregationManager.Wasm.cs - make it timer based for WASM
- split CounterGroup.Threads.cs and CounterGroup.Wasm.cs - make it timer based for WASM
- split EventPipeEventDispatcher.Threads.cs and EventPipeEventDispatcher.Wasm.cs - make it Task based
- add switch System.Net.Http.EnableMetrics and use it in HTTP client to make MetricsHandler optional
-

Api
- remove UnsupportedOSPlatform("browser")] on diagnostic API

Tests
- enable WasmPerfTracing, EventSourceSupport, MetricsSupport for unit tests which need it
- fix ActiveIssue https://github.com/dotnet/runtime/issues/93754
- enable TestSubscriptionManagerDisposal
- fix nullable argument in BeginInstrumentReporting, EndInstrumentReporting and InstrumentPublished
- enable MetricEventSourceTests, RuntimeMetricsTests, ActivityTracking

pavelsavara committed:
fix

MihaZupan commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs:
```diff
@@ -531,8 +531,11 @@ private HttpMessageHandlerStage SetupHandlerChain()
+            if (MetricsHandler.IsGloballyEnabled())
+            {
+                handler = new MetricsHandler(handler, settings._meterFactory, out Meter meter);
+                settings._metrics = new SocketsHttpHandlerMetrics(meter);
```
We'd have to update other places that rely on `_metrics` being set
https://grep.app/search?f.repo=dotnet%2Fruntime&f.repo.pattern=runtime&q=._metrics

Do we have a need for the opt-out switch outside of browser scenarios?

pavelsavara replied:
> Do we have a need for the opt-out switch outside of browser scenarios?

Maybe we don't need the opt-out for functional reasons.
But in browser, most of the time, this is just dead weight and I wonder if that should be opt-in for the browser.


> We'd have to update other places that rely on `_metrics` being set.

Thanks, I should probably cover that by some unit test. Or maybe I could split this HTTP into separated PR.

pavelsavara replied:
Moved the change to https://github.com/dotnet/runtime/pull/113528 and will incorporate your feedback there later.

pavelsavara committed:
move System.Net.Http.EnableMetrics to https://github.com/dotnet/runtime/pull/113528

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

pavelsavara committed:
ActivityTracker_Instance_Enable

pavelsavara committed:
fix

pavelsavara commented on src/libraries/System.Diagnostics.Tracing/tests/BasicEventSourceTest/ActivityTracking.cs:
```diff
@@ -5,32 +5,74 @@
             Assert.Equal(Guid.Empty, EventSource.CurrentThreadActivityId);
             es.ExampleStart();
-            Assert.NotEqual(Guid.Empty, EventSource.CurrentThreadActivityId);
+            //Assert.NotEqual(Guid.Empty, EventSource.CurrentThreadActivityId);
```
revert

pavelsavara committed:
fix

pavelsavara commented on src/libraries/System.Diagnostics.Tracing/tests/BasicEventSourceTest/ActivityTracking.cs:
```diff
@@ -5,28 +5,68 @@
 namespace BasicEventSourceTests
 {
-    public class ActivityTracking
+    public class ActivityTracking : IDisposable
```
This passed https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-23a8882c7437454e98/WasmTestOnChrome-ST-System.Diagnostics.Tracing.Tests/1/xharness-output/testResults.xml?helixlogtype=result

pavelsavara commented on src/libraries/System.Diagnostics.DiagnosticSource/tests/RuntimeMetricsTests.cs:
```diff
@@ -158,7 +158,7 @@ public void CpuTime()
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))]
+        [Fact]
```
passe https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-23a8882c7437454e98/WasmTestOnChrome-ST-System.Diagnostics.DiagnosticSource.Tests/1/xharness-output/testResults.xml?helixlogtype=result

pavelsavara commented on src/libraries/Microsoft.Extensions.Diagnostics/tests/Microsoft.Extensions.Diagnostics.Tests.csproj:
```diff
@@ -6,6 +6,11 @@
     <IncludeRemoteExecutor>true</IncludeRemoteExecutor>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetOS)' == 'browser'">
```
passing https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-23a8882c7437454e98/WasmTestOnChrome-ST-Microsoft.Extensions.Diagnostics.Tests/1/xharness-output/testResults.xml?helixlogtype=result

pavelsavara committed:
more tests

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

pavelsavara commented on src/libraries/System.Diagnostics.Tracing/tests/System.Diagnostics.Tracing.Tests.csproj:
```diff
@@ -6,6 +6,12 @@
     <TestDisableParallelization>true</TestDisableParallelization>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(TargetOS)' == 'browser'">
+    <MetricsSupport>true</MetricsSupport>
+    <EventSourceSupport>true</EventSourceSupport>
```
Because they are disabled by default

https://github.com/dotnet/sdk/blob/b9453f171a1bb911913875a54b282e87d18547af/src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.6_0.targets#L45

Maybe this is not the only place.

And I think we should disable `MetricsSupport` as well.

maraf replied:
That is Blazor SDK and these are runtime tests. Blazor targets are not imported here.

pavelsavara commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.Wasm.cs:
```diff
@@ -0,0 +1,50 @@
+            Debug.Assert(Monitor.IsEntered(m_dispatchControlLock));
+            Debug.Assert(sessionID != 0);
+
+            m_dispatchTaskCancellationSource = new CancellationTokenSource();
```
Yes, `DisableEvents()` and `EnableEvents()` -> `CommitDispatchConfiguration()` -> `SetStopDispatchTask()`

pavelsavara committed:
fix

pavelsavara committed:
fix

javiercn commented on src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj:
```diff
@@ -48,6 +50,9 @@
     <NodeNpmModule Include="node-fetch" />
     <NodeNpmModule Include="node-abort-controller" />
 
+    <RuntimeHostConfigurationOption Include="System.Net.Http.EnableActivityPropagation" Value="true" Trim="true" />
+    <RuntimeHostConfigurationOption Include="System.Net.Http.EnableMetrics" Value="true" Trim="true" />
```
@pavelsavara do we have another case where this won't work during the dev loop? (at least for now)

pavelsavara replied:
Indeed, same https://github.com/dotnet/runtime/issues/97449

pavelsavara replied:
Only blazor dev loop. The non-blazor WASM should work I think.

maraf replied:
> Only blazor dev loop. The non-blazor WASM should work I think.

Only runtime tests are loading the runtime config.bin

javiercn commented on src/libraries/System.Diagnostics.DiagnosticSource/tests/MetricEventSourceTests.cs:
```diff
@@ -2313,12 +2314,24 @@ void WaitForEvent(TimeSpan timeout, int numEvents, string eventName)
+                {
+                    // in the single-threaded browser environment, we need to yield to the browser to allow the event to be processed
+                    // we also can't block with WaitOne
+                    await Task.Delay(10);
```
```suggestion
                    await Task.Yield();
```
Or is there a reason to delay 10ms

pavelsavara replied:
This is `OuterLoop` test and we are inside `while` loop which is waiting. The propagation is on 100ms timer. So we wait up to 10 ticks. We could possibly use some non-blocking signal, like `Task` to make it better. But it's not worth the effort.

javiercn replied:
Sounds good.

The other thing that I was going to suggest is using a TCS. We do that in tests when we want to make sure they aren't flaky (as you get to control when it gets triggered)

javiercn commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.Wasm.cs:
```diff
@@ -0,0 +1,50 @@
+
+            m_dispatchTaskCancellationSource = new CancellationTokenSource();
+            Task? previousDispatchTask = m_dispatchTask;
+            if (previousDispatchTask != null)
```
Could you initialize this to `Task.CompletedTask` and avoid the `if`?

javiercn commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.Wasm.cs:
```diff
@@ -0,0 +1,50 @@
+
+            // Wait for SignalSession() to be called before we call disable, otherwise
+            // the SignalSession() call could be on a disabled session.
+            await Task.Delay(1, CancellationToken.None).ConfigureAwait(false);
```
```suggestion
            await Task.Yield(CancellationToken.None).ConfigureAwait(false);
```

Can this be `Task.Yield`? I'm assuming your intent is to just yield the thread?

pavelsavara replied:
Let's try

jkotas commented:
cc @dotnet/area-system-diagnostics-metric @dotnet/area-system-diagnostics-tracing

jkotas commented:
@ericstj This is introducing RID-specific binary in System.Diagnostics.DiagnosticSource nuget package. Can it lead to issues or less than ideal behaviors?

pavelsavara committed:
Update src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.Wasm.cs

Co-authored-by: Javier Calvarro Nelson <jacalvar@microsoft.com>

pavelsavara committed:
fix

tarekgh commented:
/azp run runtime-libraries-coreclr outerloop

tarekgh commented:
CC @noahfalk

pavelsavara committed:
fix

pavelsavara commented:
/azp run runtime-libraries-coreclr outerloop

jkotas commented on src/libraries/System.Diagnostics.DiagnosticSource/tests/System.Diagnostics.DiagnosticSource.Tests.csproj:
```diff
@@ -5,9 +5,14 @@
     <IncludeRemoteExecutor>true</IncludeRemoteExecutor>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetOS)' == 'browser'">
```
Should this be unconditional? Otherwise, we may want to add extra target to `TargetFrameworks`.

pavelsavara replied:
I think that on other platforms, those are not need because they are on by default.
But I see this as relevant to particular CI leg run, not that we need another flavor of the DLL.

I'm usually confused about those, but I don't what's wrong with that mental model. Hints ? 
Thanks.

jkotas replied:
The libraries build expects each configuration to produce separate binary. There is `-allConfigurations` command line options that builds all configurations in one build script invocation. The general rule is that the conditions in the project file need to match `TargetFrameworks`.

pavelsavara replied:
I think there is no harm by enabling those flag unconditionally for this test project.

ViktorHofer commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
+    <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppPrevious);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum);$(NetCoreAppCurrent)-browser</TargetFrameworks>
```
Our recommendation is that libraries that also ship as nuget package shouldn't be RID specific. What hinders this code base form using runtime conditions?

pavelsavara replied:
I think I can re-work this to be behind `OperatingSystem.IsBrowser()` and few places where I disable warnings about `new Thread`. 

But I would like to understand more, can you please explain ?
I thought this is inside browser runtime pack, which is completely platform specific in browser OS.

ViktorHofer replied:
> I thought this is inside browser runtime pack, which is completely platform specific in browser OS.

This project also ships as a standalone nuget package: https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/10.0.0-preview.1.25080.5#readme-body-tab

For libraries that ship standalone as a nuget package or in both the shared framework and via a package the above recommendation applies. Libraries that only ship via the shared framework are free to target specific RIDs (even though the poor developer innerloop in runtime applies there as well).

pavelsavara replied:
Thank you for patience with me @ViktorHofer 

> poor developer innerloop

What do you mean by that ? Is that because the DLL of different flavor behaves differently ? If I make this dynamic condition, it would behave differently as well. Or something else ?

> package size

How could be platform specific DLL larger than one which contains all code paths ?

pavelsavara replied:
Or you mean NuGet size ?
(which we care much less about than deployed browser download size)

jkotas replied:
> deployed browser app 

If you trim the deployed app, multiple binaries vs. conditions should not have a meaningful difference. The trimmer should be able to recognize the `IsBrowser()` conditions and eliminate the unrechable code.
If you do not trim the deployed app, the extra code under dynamic conditions is a drop in a bucket.

ericstj replied:
An additional wrinkle is that runtime-packages *require* deps file resolution to select the best runtime assets.  Sometimes library projects will be deployed for use in a system (eg: plugins, roslyn analyzers, etc) that does not do deps resolution so introducing runtimes into an existing package will create friction here.

pavelsavara replied:
done

pavelsavara commented on src/libraries/Microsoft.Extensions.Diagnostics/tests/Microsoft.Extensions.Diagnostics.Tests.csproj:
```diff
@@ -6,6 +6,11 @@
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetOS)' == 'browser'">
+    <MetricsSupport>true</MetricsSupport>
```
@ViktorHofer why do we have those `<RuntimeHostConfigurationOption` items in the SDK and not in the runtime repo ? 
Does that also complicate versioning of the default ?

(I don't plan to do anything about it in this PR, just asking)

tarekgh replied:
`MetricsSupport` is a config switch to allow the aot linker to trim the whole metrics code when it is set to `false`. This switch is set to `false` for wasm and MAUI too by default. 

> Like maybe adding // MetricsSupport to

I am not getting what you mean by that? the config switch can be enabled by the app and not by the library.

mdh1418 replied:
My only concern is that the config switch is not obvious from dotnet/runtime code itself. `MetricsSupport` isn't consumed from dotnet/runtime itself (the consumer is in the sdk side right?), and there isn't as much usage/docs in dotnet/runtime compared to `EventSourceSupport`, so someone in the future might come across this property and remove it without knowing that its actually being used.

mdh1418 replied:
I was just thinking of leaving some note to associate what `MetricsSupport` does, such as a comment `// MetricsSupport` by where `System.Diagnostics.Metrics.Meter.IsSupported` is, there would at least be a connection between `<MetricsSupport>true</MetricsSupport>` and an "effect" in the runtime.

Analternative would be some doc in runtime saying something like `<MetricsSupport>true</MetricsSupport> is a config switch that ...`

tarekgh replied:
Ah, comment is good. We have it documented in https://github.com/dotnet/docs/blob/main/docs/core/deploying/trimming/trimming-options.md#trim-framework-library-features so maybe in the comment we can point at this doc.

jkotas replied:
> why do we have those <RuntimeHostConfigurationOption items in the SDK and not in the runtime repo ?

These items need to end up as msbuild targets in the SDK. If we had them in runtime, we would have to get them into SDK via transport package or something like that. It would be different, but not necessarily simpler.

pavelsavara replied:
added comment and link

mdh1418 commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventDispatcher.Wasm.cs:
Just curious, has there been other perf testing of EventPipe for wasm besides the 5 tests running in the `browser-wasm linux Release AllSubsets_Mono_RuntimeTests monointerpreter` [pipeline](https://dev.azure.com/dnceng-public/public/_build/results?buildId=984544&view=logs&j=3e93a48f-6568-53f2-76b3-1285c183c39d&t=dbc6652c-922c-520b-2284-9f34635f43ea)? (One of the tests `name_config_with_pid` is an auto-pass because its actually not supported and the other 4 (eventcounter/regression-* and eventlistener/eventlistener*) seem like very light-weight tests.

I believe we've seen a few issues pop up with respect to performance of EventPipe, where the in-proc NativeRuntimeEventSource spends alot of CPU time on EventPipeInternal.GetNextEvent (EventPipeEventDispatcher.cs).

pavelsavara replied:
At this time, I just trying to make it work at all. The perf would be interesting problem for sure, since this is happening on the same thread as the app.

noahfalk replied:
Fwiw, the perf issue we've seen in the past is that the DispatchEventsToEventListenersOnce loop can get hot, spending most of its time inside the call to EventPipeInternal.GetNextEvent(). The native implementation backing that call is inefficiently doing a time-ordered merge sort of the events from multiple threads. The loop overall has O(N*M) performance for an app with N threads emitting M total events. The constant factor may be mediocre as well. The browser might benefit by its lack of threads, or it might appear worse because the dispatch thread is directly competing for CPU time with app's foreground work.

noahfalk commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.Wasm.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace System.Diagnostics.Tracing
+{
+    internal sealed partial class CounterGroup
```
Is there something specific you need from EventCounters or you are adding it for completeness? I view EventCounter as an outdated approach that has been superceded by System.Diagnostics.Metrics. We still support it for back-compat reasons, but we don't encourage it for new scenarios. Since browser has never supported the scenario you have the opportunity to skip it and just focus your investments on S.D.M.

https://learn.microsoft.com/en-us/dotnet/core/diagnostics/compare-metric-apis

pavelsavara replied:
I'm trying to get on par with other platforms.

We will be adding new Metrics for Blazor as later step.

noahfalk replied:
> I'm trying to get on par with other platforms.

Have we seen any complaints from customers that EventCounters don't work on browser? I wasn't aware of much demand there. You will be taking on an extra support burden (https://github.com/dotnet/runtime/issues?q=is%3Aissue%20EventCounters%20) as soon as you enable them. This also adds more code to scenarios that I assume are code-size sensitive.

> We will be adding new Metrics for Blazor as later step.

Using the S.D.M API or the EventCounters API? I'd recommend using the S.D.M API.

pavelsavara replied:
> Have we seen any complaints from customers that EventCounters don't work on browser?

At the moment none of diagnostics works in the browser at all AFAIK, so there is nothing to complain about.
There is no event pipe, now way how to get your events. 

This PR and also upcoming EP/DS is part of https://github.com/dotnet/runtime/issues/76316

The issue has nice 29 upvotes.

> > We will be adding new Metrics for Blazor as later step.

> Using the S.D.M API or the EventCounters API? I'd recommend using the S.D.M API.

S.D.M. 

I will get back to you when we have some draft/design of what kind of metrics make sense for Blazor and ask for advice.

pavelsavara replied:
It's good to know where scope could be cut eventually. 
Do you propose any of this should be made experimental or opt-in ? Also for perf reasons.

noahfalk replied:
For the EventCounter part I'd suggest not including DiagnosticCounter or any of its derived classes. After that it should make CounterGroup unreachable.

EventSource and S.D.M already have their build time opt-ins. EventSource supports egressing events as ETW manifest events, ETW TraceLogging events, EventPipe events, or EventListener events. The code paths are somewhat intertwined right now but if you expect devs want to publish profiling enabled versions of their websites to users (production perf testing?) maybe it would be worthwhile to opt into those pathways individually to keep the code-size overhead lower? This feels like an optimization you could decide on later rather than something I'd suggest you go after right away.

pavelsavara replied:
> if you expect devs want to publish profiling enabled versions of their websites to users (production perf testing?) 

Yes, for example ActivityTracking doesn't work without EP/DS native module linked in.
I'm introducing `<WasmPerfTracing>` which does that. 

From managed code perspective I'm enabling `<FeaturePerfTracing>` for browser. So some size cost is already there.

> maybe it would be worthwhile to opt into those pathways individually 
> ETW TraceLogging events, 
> EventPipe events, or EventListener events. 

I prefer functional parity over size optimization, when `<EventSourceSupport>` is on.

I would add more opt-out granularity only when some of those paths overwhelm other more useful scenarios with perf overhead.

> For the EventCounter part I'd suggest not including DiagnosticCounter

This is because it's kind of obsolete and we prefer that people use S.D.M instead, right ?
But from functional perspective my implementation works. 
So the question is really if by not adding this I'm limiting some legacy code to be used in browser as is.

Anyway, I can move DiagnosticCounter to separate PR and we can negotiate it there.

pavelsavara replied:
moved to https://github.com/dotnet/runtime/pull/113682

pavelsavara replied:
I guess I need to take that back because `RuntimeEventSource` uses lot of `DiagnosticCounter` internally.

noahfalk replied:
Sure thing.

> This is because it's kind of obsolete and we prefer that people use S.D.M instead, right ?

Yeah, and to save work in the future. As soon as it is supported then it needs tests, investigating when those tests fail, customers might file bugs against it, etc. Its nice not to open that door if we don't expect it to be very valuable.

> So the question is really if by not adding this I'm limiting some legacy code to be used in browser as is.

Its possible, but if that were true I'd expect you would have already heard about it. Instead of customers complaining that they can't profile I assume they'd be complaining their code didn't port.

noahfalk replied:
My suggestion is not to use EventCounters internally either. You could disable EventSources which use it (or include the APIs and no-op the EventCounter implementation if this is too hasslesome).
RuntimeEventSource has a superceding set of S.D.M based metrics available here: [RuntimeMetrics.cs](https://source.dot.net/#System.Diagnostics.DiagnosticSource/System/Diagnostics/Metrics/RuntimeMetrics.cs,10)

pavelsavara replied:
For the `dotnet-counters` it means that `System.Runtime` provider would not work, but the `System.Diagnostics.Metrics` provider with `Metrics=System.Runtime` is also requested by default, right ?
So, there would be no functional gap ?

pavelsavara replied:
> My suggestion is not to use EventCounters internally either. 

done

pavelsavara commented on src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MetricsEventSource.cs:
```diff
@@ -173,7 +173,7 @@ public void BeginInstrumentReporting(
         {
             WriteEvent(7, sessionId, meterName, meterVersion ?? "", instrumentName, instrumentType, unit ?? "", description ?? "",
-                    instrumentTags, meterTags, meterScopeHash, instrumentId, meterTelemetrySchemaUrl);
+                    instrumentTags, meterTags, meterScopeHash, instrumentId, meterTelemetrySchemaUrl ?? "");
```
It's bug I discovered when testing this. It would fail to create the event if something passed the null

pavelsavara replied:
Do you want me to create separate PR for this so that it could be backported ?

noahfalk replied:
SchemaUrl was only added recently in Dev10 and I don't think we'd care about servicing a preview release so no need. Coincidentally I think we just discovered the same bug independently in some of our metrics tests for the dotnet-counters tool. The libraries tests for this only validate the events through EventListener which is more lenient about the null.

fyi @tarekgh

pavelsavara commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.Threads.cs:
```diff
@@ -0,0 +1,54 @@
+            {
+                s_pollingThreadSleepEvent = new AutoResetEvent(false);
+                s_counterGroupEnabledList = new List<CounterGroup>();
+                // TODO
```
I will delete it. Thanks

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

pavelsavara committed:
feedback

- don't split System.Diagnostics.DiagnosticSource into separate platform DLL
- add comments for MetricsSupport and EventSourceSupport features in tests, make them global

pavelsavara committed:
feedbak - keep DiagnosticCounter not supported API

pavelsavara committed:
more

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
fix

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

pavelsavara committed:
feedback

pavelsavara committed:
fix

pavelsavara committed:
just browser

pavelsavara commented:
/azp run runtime-libraries-coreclr outerloop

pavelsavara commented:
@noahfalk I believe I processed all your feedback. Please have another look.

@tarekgh I executed coreCLR outerloop pipeline and there are quite few timeouts I'm not familiar with.
It seems to me most of them are completely unrelated, because I think I made no change to CoreCLR.

Some of them are about `System.Diagnostics.DiagnosticSource.Tests`, 
**passing and timeout**
https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-4ce231f817134eeeb1/System.Diagnostics.DiagnosticSource.Tests/1/console.97f0f45e.log?helixlogtype=result
https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-d64bee06b6904cb984/System.Diagnostics.DiagnosticSource.Tests/1/console.b25b5eaf.log?helixlogtype=result

**timeout during test**
https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-8849796b28c94bb8af/System.Diagnostics.DiagnosticSource.Tests/1/console.e5a2c1aa.log?helixlogtype=result
https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-1db6ed0772b4439aae/System.Diagnostics.DiagnosticSource.Tests/1/console.6866e220.log?helixlogtype=result

If you guys believe that my changes in `MetricEventSourceTests.cs` to make it async are causing that I could revert the changes in that file and exclude the test on browser. It's outerloop anyway.

please advise, thanks

tarekgh commented:
> It seems to me most of them are completely unrelated, because I think I made no change to CoreCLR.

I am seeing this timeout https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-113524-merge-1db6ed0772b4439aae/System.Diagnostics.DiagnosticSource.Tests/1/console.6866e220.log?helixlogtype=result, looks these failing in slow platforms (like Debug and Linux Musl). These outer loop tests take a long time so on slow platforms can timeout I guess. I don't recall seeing such failures, but I can see why it can time out. I have re-run the failing tests hopefully it passes. If it is still failing, you may ignore them for now as it is time out and passing in other platforms.

noahfalk reviewed (approved):
👍

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

# Conflicts:
#	src/libraries/Microsoft.Extensions.Diagnostics/tests/Microsoft.Extensions.Diagnostics.Tests.csproj

pavelsavara committed:
disable RequestDuration_EnrichmentHandler_ContentLengthError_Recorded for firefox

pavelsavara committed:
Merge branch 'main' into browser_managed_diag_events

pavelsavara committed:
RequestDuration_Success_Recorded not firefox

pavelsavara commented:
/ba-g CI issues are known
